pep-0725.rst

```diff
@@ -0,0 +1,236 @@
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 28-Jul-2023
+Python-Version: 3.10
+Post-History: 
+Resolution: 
+
+
+Abstract
+========
+
+:pep:`647` introduced the concept of ``TypeGuard`` functions which return true
+if their input parameter matches their target type. For example, a function that
+returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's
+input parameter is a ``str``. This allows type checkers to narrow types in this
+positive case.
+
+This PEP further refines :pep:`647` by allowing type checkers to also narrow types
+when a ``TypeGuard`` function returns false.
+
+Motivation
+==========
+
+``TypeGuards`` are used throughout Python libraries to allow a type checker to
+narrow the type of something when the ``TypeGuard`` returns true.
+
+However, in the ``else`` clause, :pep:`647` didn't prescribe what the type might
+be:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]:
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type here is not narrowed. It is still 'str | int' 
+
+
+This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype
+of the type of the first input parameter, then the false return case can be
+further narrowed. 
+
+This changes the example above like so:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]:
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type checkers can assume val is an 'int' in this branch
+
+Since the ``TypeGuard`` type (or output type) is a subtype of the first input
+parameter type, a type checker can determine that the only possible type in the
+``else`` is the other type in the ``Union``. In this example, it is safe to
+assume that if ``is_str`` returns false, then type of the ``val`` argument is an
+``int``.
+
+
+Specification
+=============
+
+This PEP requires no new changes to the language. It is merely modifying the
+definition of ``TypeGuard`` for type checkers. Runtimes are already behaving
+in this way.
+
+
+Existing ``TypeGuard`` usage may change though, as described in 
+`Backwards Compatibility`_.
+
+Unsafe Narrowing
+--------------------
+
+There are cases where this further type narrowing is not possible though. Here's
+an example:
+
+.. code-block:: python
+
+    def is_str_list(val: list[int | str]) -> TypeGuard[list[str]]
+        return all(isinstance(x, str) for x in val)
+
+    def func(val: list[int | str]):
+        if is_str_list(val):
+            # Type checker assumes list[str] here
+        else:
+            # Type checker cannot assume list[int] here
+
+Since ``list`` is invariant, it doesn't have any subtypes. ``list[str]`` is not
```

Comment:
I think this misses the point. If you used Sequence instead of list then you have a covariant type, but you still wouldn't able to narrow in the negative case.

Reply:
I was trying for an example that would obviously not be a subtype of the input. Do you think a different example would be better?

Something like:

```python

class A():
    pass

class B(A):
    pass

def is_a(x: A | B) -> TypeGuard[A]:
    return isinstance(x, A)
```